branches/0.9.0/kxneur/kxneur.1
.TH "kxneur" "1" 
.SH "NAME" 
kxneur \(em KDE frontend for xneur keyboard layout switcher 
.SH "SYNOPSIS" 
.PP 
\fBkxneur\fR [\fBQt-options\fP]  [\fBKDE-options\fP]  
.SH "DESCRIPTION" 
.PP 
This manual page documents briefly \fBkxneur\fR command. 
.PP 
kXNeur is a KDE frontend for xneur keyboard layout switcher. kXNeur runs in system tray and shows XNeur's state. It also allows to configure XNeur via GUI dialog. 
.SH "OPTIONS" 
.PP 
Thes program accepts the same options any usual KDE appilcations do. 
.PP 
Generic options: 
.IP "\fB\-\-help\fP" 10 
Show help about options 
.IP "\fB\-\-help-qt\fP" 10 
Show Qt specific options 
.IP "\fB\-\-help-kde\fP" 10 
Show KDE specific options 
.IP "\fB\-\-help-all\fP" 10 
Show all options 
.IP "\fB\-\-author\fP" 10 
Show author information 
.IP "\fB-v\fP, \fB\-\-version\fP" 10 
Show version information 
.IP "\fB\-\-license\fP" 10 
Show license information 
.IP "\fB\-\-\fP" 10 
End of options 
.PP 
Qt options: 
.IP "\fB\-\-display <displayname>\fP" 10 
Use the X-server display 'displayname' 
.IP "\fB\-\-session <sessionId>\fP" 10 
Restore the application for the given 'sessionId' 
.IP "\fB\-\-cmap\fP" 10 
Causes the application to install a private color map on an 8-bit display 
.IP "\fB\-\-ncols <count>\fP" 10 
Limits the number of colors allocated in the color cube on an 8-bit display, if the  application  is  using  the  QApplication::ManyColor color specification 
.IP "\fB\-\-nograb\fP" 10 
tells Qt to never grab the mouse or the keyboard 
.IP "\fB\-\-dograb\fP" 10 
running under a debugger can cause an implicit \-nograb, use \-dograb to override 
.IP "\fB\-\-sync\fP" 10 
switches to synchronous mode for debugging 
.IP "\fB\-\-fn\fP, \fB\-\-font <fontname>\fP" 10 
defines the application font 
.IP "\fB\-\-bg\fP, \fB\-\-background <color>\fP" 10 
sets the default background color and an application palette (light and dark shades are calculated) 
.IP "\fB\-\-fg\fP, \fB\-\-foreground <color>\fP" 10 
sets the default foreground color 
.IP "\fB\-\-btn\fP, \fB\-\-button <color>\fP" 10 
sets the default button color 
.IP "\fB\-\-name <name>\fP" 10 
sets the application name 
.IP "\fB\-\-title <title>\fP" 10 
sets the application title (caption) 
.IP "\fB\-\-visual TrueColor\fP" 10 
forces the application to use a TrueColor visual on an 8-bit display 
.IP "\fB\-\-inputstyle <inputstyle>\fP" 10 
sets XIM (X Input Method) input style. Possible values are onthespot, overthespot, offthespot and root 
.IP "\fB\-\-im <XIM server>\fP" 10 
set XIM server 
.IP "\fB\-\-noxim\fP" 10 
disable XIM 
.IP "\fB\-\-reverse\fP" 10 
mirrors the whole layout of widgets 
.PP 
KDE options: 
.IP "\fB\-\-caption <caption>\fP" 10 
Use 'caption' as name in the titlebar 
.IP "\fB\-\-icon <icon>\fP" 10 
Use 'icon' as the application icon 
.IP "\fB\-\-miniicon <icon>\fP" 10 
Use 'icon' as the icon in the titlebar 
.IP "\fB\-\-config <filename>\fP" 10 
Use alternative configuration file 
.IP "\fB\-\-dcopserver <server>\fP" 10 
Use the DCOP Server specified by 'server' 
.IP "\fB\-\-nocrashhandler\fP" 10 
Disable crash handler, to get core dumps 
.IP "\fB\-\-waitforwm\fP" 10 
Waits for a WM_NET compatible windowmanager 
.IP "\fB\-\-style <style>\fP" 10 
sets the application GUI style 
.IP "\fB\-\-geometry <geometry>\fP" 10 
sets the client geometry of the main widget \- see man X for the argument format 
.SH "SEE ALSO" 
.PP 
xneur (1), gxneur(1) 
.SH "AUTHOR" 
.PP 
This manual page was written by Alexander GQ Gerasiov <gq@cs.msu.su> for 
the \fBDebian\fP system (but may be used by others).  Permission is 
granted to copy, distribute and/or modify this document under 
the terms of the GNU General Public License, Version 2 any  
later version published by the Free Software Foundation. 
 
.PP 
On Debian systems, the complete text of the GNU General Public 
License can be found in /usr/share/common-licenses/GPL. 
 
.\" created by instant / docbook-to-man, Thu 27 Dec 2007, 01:05 


xneur/NEWS
October 24 2013:
Version 0.17.0 released.
Add misprint correction
Add total cancellation of correction
Check memory leaks via сppcheck
Add new option "Rotate autocompletion"
Add app list for exclude from KeyRelease sending
Big lot of bugfixes (see mercurial journal).

October 25 2012:
Version 0.16.0 released.
New option for сorrect two minus with a long dash
New option for сorrect ... with a …
Add multistring abbreviation
Remove deprecated Xlib function
Add support window with WM_NAME atom
Big lot of bugfixes (see mercurial journal).

November 10 2011:
Version 0.15.0 released.
New option for сorrect two minus with a dash
New option for сorrect (c) with a copyright sign
New option for сorrect (tm) with a trademark sign
New option for сorrect (r) with a registered sign

October 24 2011:
Version 0.14.0 released.
New troubleshooting option for disable mouse tracking.
New troubleshooting option for fullscreen apps.
New option for сorrect two space with a comma and a space
Bugfix for change case action.
Bugfix for Gtk3 apps.

May 24 2011:
Version 0.13.0 released.
Enabled hotkeys on excluded apps.
Add cash for keyboard map. 
New option - monitor the input.
Big lot of bugfixes (see svn journal).

February 01 2011:
Version 0.12.0 released.
New troubleshooth option - compatibility with completion.
New troubleshooth option - flush buffer when pressed Escape.
New option - correction small letter to capital letter after dot.
New hotkey - inserd date.
Big lot of bugfixes (see svn journal).

November 26 2010:
Version 0.11.0 released.
Using Enchant Library.
New file format for user dictionaries.
Add OpenMP for cicles.
Add new keyboard binds - block keyboard, rotate layout after cwith selected text, back rotate layouts.
Lot of bugfixes.

October 07 2010:
Version 0.10.0 released.
Bugfix for locales (bug in XLookupString)
Send archived log to e-mail possile to named server (smtp.server.org, ex.)
Log writed to html.

May 21 2010:
Version 0.9.9 released.
Bugfix for change layots count where xneur run.
Bugfix for new version D-Bus.
Bugfix for modificator keys.
Bugfix for ISO Level 3 keys.

March 19 2010:
Version 0.9.8 released.
Automatic identification of system keyboard layouts.
Calculating clipboard and selected formul's via calc deleted.
New options for troubleshooting.
New library (libxneur) for using xneur functional in external apps.
Add georgian language.
Lot of bugfixes

October 14 2009:
Version 0.9.7 released.
Add option for preview conversation of selected text.
Add volume level option.
Add option for individual disabling оf notification.
Bug fix for linked library

September 16 2009:
Version 0.9.6 released.
Add option to send archived log to e-mail
Add plugins system and plugin for xneur statistic.
Calculating clipboard and selected formul's via calc.
Bug fix for numeric keyboard

August 04 2009:
Version 0.9.5 released.
Add lock to Caps
Add check puntuation
Pattern mining and recognition
Clipboard correction
Add popup messages

April 21 2009:
Version 0.9.4 released.
Remove pixmap flag near of cursor
Add Bulgarian, Czech, Greek, Estonian, Armenian, Lithuanian, Latvian, Poland, Spanish and Uzbek languages
Transliteration ISO 9 (GOST 7.79-2000) 
New algo for check language (optional)
Add option for layout fix
Bug fix for diacritical sign

January 09 2009:
Version 0.9.3 released.
Add OSD support
New image library (imlib2) support
i18n support in xneur

November 02 2008
Version 0.9.2 released.
KDE application events receive bugfix
Cancel UTF checking
Add abbreviations replacement function
Fix Wine and SDL apps processing
Add new features:
	correction incidental caps
	correction two capital letter
	flush internal buffer when press Enter and Tab
	don't process when press Enter and Tab
Lot of bugfixes

July 23 2008
Version 0.9.1 released.
KDE application events receive bugfix
Sounds fixes
Lot of bugfixes

June 24 2008
Version 0.9.0 released.
Rewrite event receiving and sending
Fixed mouse event support
Add pixmap flag beside of the mouse cursor
Add keyboard logging to file
Changed man pages
Lot of bugfixes

October 11 2007
Version 0.8.0 released.
Add sound support for XNeur events.
Add aspell support for the better detection of language.
Add sound support.
Add German and Kazakh languages.
Rewrite keymap support.
Changed configuration script.
Add man pages.

July 17 2007
Version 0.6.2 released.

May 17 2007:
Version 0.6.1 released.
Many fixes in configure scripts.
Add support old versions Xorg.
Fix bug with new desktop.
New mechanism for OS standart signals.

April 24 2007:
Version 0.6 released.
Add new language.
Some optimize, bug fixes.
Fix Beryl, Compiz, etc errors.
GUI changes.

March 10 2007:
Version 0.5 released.
Add new language and regular expression mechanism.

January 23 2007:
Version 0.4 released.
New options and new algorithms.

December 20 2006:
Version 0.3 released.
This release much more user-friendly :)

November 15 2006:
Version 0.2 released.

June 9 2005:
New tree is opened, i decide to remake structure,
code style and others for good code reading and understanding 
and for better bug fixing.

Apr 18 2005:
I decide to upload cvs snapshot before 0.0.4 release,
so it's be a good news for users, some features avialable now,
some fixes, include memory leaks fixing.
Some optimize, but I find some errors on xlib part ... 
it's hard to me, becuase the following troubles occured -
 - it's not my code and i need to rewite it all
 - i don't known XLib too good for makeing it faster
please understand me - i have no time for this project now, 
i need to quickly education on source code and XLib functions.
But I hope that in some days i will be have a really stable 
version of it project.
and for this purpose i need for good testing - please help me with it.
thanx.

Jan 20 2005:
Third testing release is out!
Fixed a lot of erros and bugs in X part of project,
added documentation (not all), have started 
writing AI like algorithm.

Jan 05 2005:
Testing 0.0.2 release came out!

Nov 08 2004:

kxneur/src/kxneur.cpp
/**********************************************************************************
 *  KXNeur (KDE X Neural Switcher) is XNeur front-end for KDE ( http://xneur.ru ).*
 *  Copyright (C) 2007-2008  Vadim Likhota <vadim-lvv@yandex.ru>                  *
 *                                                                                *
 *  This program is free software; you can redistribute it and/or modify          *
 *  it under the terms of the GNU General Public License as published by          *
 *  the Free Software Foundation; either version 2 of the License, or             *
 *  (at your option) any later version.                                           *
 *                                                                                *
 *  This program is distributed in the hope that it will be useful,               *
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of                *
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                 *
 *  GNU General Public License for more details.                                  *
 *                                                                                *
 *  You should have received a copy of the GNU General Public License             *
 *  along with this program; if not, write to the Free Software                   *
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA    *
 **********************************************************************************/

#include <signal.h>
#include <kaction.h>
#include <kpopupmenu.h>
#include <qstringlist.h>
#include <kstandarddirs.h>
#include <kiconloader.h>
#include <qpainter.h>
#include <kconfig.h>
#include <klocale.h>
#include <qmessagebox.h>
#include "kxneur.h"
#include "kxneurtray.h"
#include "kxneursettings.h"
#include "kxneurconf.h"
#include "kxnkeyboard.h"


KXNeurApp::KXNeurApp()
#ifdef XN_END
 : KUniqueApplication()
#else
 : KApplication()
#endif
{
    KGlobal::dirs()->addResourceDir("appdata", ".");
    all_langs = new KConfig("langs", true, true, "appdata");
    all_langs->setGroup("Languages");

    xnkb = KXNKeyboard::self();
    trayicon = new KXNeurTray();
    setMainWidget( trayicon );
    trayicon->show();

    // xneur_pid = 0;
    xnconf = NULL;
    cur_lang = -1;
    prev_lang = -1;
    for ( int i = 0 ; i < MAX_LANGUAGES ; i++ )
	langs[i] = NULL;

    if ( !(KXNeurSettings::self()->RunXNeur() && xneur_start()) ) {
#ifndef XN_END
	qDebug("start -- ok\n");
#endif
	trayicon->run->setText(i18n("Start xneur daemon"));
	trayicon->run->setIcon("fork");
	trayicon->mode->setEnabled(false);
	// xnconf_reload();
    }
    cur_lang = xnkb->getGroupNo();
    refreshLang();
    trayicon->setPixmap(langs[cur_lang]->pic);

    QObject::connect(xnkb, SIGNAL(groupChanged(int)), this, SLOT(groupChange(int)));
    QObject::connect(xnkb, SIGNAL(layoutChanged()), this, SLOT(refreshLang()));
    QObject::connect(trayicon, SIGNAL(clicked()), this, SLOT(setNextLang()));
    QObject::connect(trayicon, SIGNAL(quitSelected()), this, SLOT(slotExit()));
}


KXNeurApp::~KXNeurApp()
{
    xneur_stop();
}


bool KXNeurApp::xneur_start()
{
    int major_v, minor_v;

    if ( xnconf )
	if (!xneur_stop()) {
#ifndef XN_END
	    qDebug("start -- fail 1\n");
#endif
	    return false;
	}
	else
	    xnconf = NULL;
    xnconf = xneur_config_init();
    if ( xnconf > 0 ) {
	xnconf->get_library_version(&major_v, &minor_v);
	if ( major_v != XNEUR_MAJOR_VERSION && minor_v != XNEUR_MINOR_VERSION ) {
		qDebug("Wrong XNeur configuration library api version.\nPlease install libxnconfig version %d.x\n", XNEUR_MAJOR_VERSION);
		QMessageBox::critical(0, "XNeur", i18n("Wrong XNeur configuration library api version.\nPlease install libxnconfig version ")+
			QString::number(XNEUR_MAJOR_VERSION)+".x");
                xnconf->uninit(xnconf);
		quit();
	}
	// xnconf_reload();
	if ( !xnconf->load(xnconf) ) {
		qDebug("XNeur config broken!\nPlease, remove ~/.xneur/xneurrc and reinstall XNeur package!\n");
		QMessageBox::critical(0, "XNeur", i18n("XNeur config broken!\nPlease, remove ~/.xneur/xneurrc and reinstall XNeur package!\n"));
                xnconf->uninit(xnconf);
		quit();
	}
        QStringList args;
        if ( KXNeurSettings::self()->ForceRun() )
	    args << "-f";
	if ( kdeinitExec("xneur", args, 0, NULL, 0) == 0 ) {
	    xnconf_gets();
	    trayicon->run->setText(i18n("Stop xneur daemon"));
	    trayicon->run->setIcon("cancel"); // stop, no, remove
	    trayicon->mode->setEnabled(true);
#ifndef XN_END
	    qDebug("start -- ok\n");
#endif
	    return true;
	}
	qDebug("Couldn't start xneur\nVerify that it installed\n");
	QMessageBox::critical(0, "XNeur", i18n("Couldn't start xneur\nVerify that it installed\n"));
    }
    xnconf = NULL;
#ifndef XN_END
    qDebug("start -- fail 2\n");
#endif
    return false;
}


bool KXNeurApp::xneur_stop()
{
    if ( xnconf && xnconf->get_pid(xnconf) ) {
	if ( !xnconf->kill(xnconf) ) {
#ifndef XN_END
	    qDebug("stop -- fail 1\n");
#endif
            return false;
	}
	xnconf->uninit(xnconf);
	xnconf = NULL;
	trayicon->run->setText(i18n("Start xneur daemon"));
	trayicon->run->setIcon("fork"); // launch
	trayicon->mode->setEnabled(false);
#ifndef XN_END
	qDebug("stop -- ok\n");
#endif
	return true;
    }
#ifndef XN_END
    qDebug("stop -- fail 3\n");
#endif
    return false;
}


void KXNeurApp::slotExit()
{
    xneur_stop();
}

void KXNeurApp::slotUpdateMode()
{
    if ( xnconf->is_manual_mode(xnconf) )
	xnconf->set_manual_mode(xnconf, 0);
    else
	xnconf->set_manual_mode(xnconf, 1);

    xnconf_gets();
}


void KXNeurApp::xnconf_gets()
{
    cnt_langs = xnconf->total_languages;


    if ( xnconf->manual_mode ) {
	trayicon->mode->setText(i18n("Set auto mode"));
	trayicon->mode->setIcon("exec"); // mics, gear
	// printf("auto\n");
    }
    else {
	trayicon->mode->setText(i18n("Set manual mode"));
	trayicon->mode->setIcon("embedjs");
	// printf("manual\n");
    }
}


void KXNeurApp::slotUpdateRun()
{
    if ( xnconf )
	xneur_stop();
    else
	xneur_start();
}


void KXNeurApp::slotPref()
{
    KXNeurConf *dlg = new KXNeurConf(this);

    dlg->exec();

    delete dlg;

    refreshLang();
    trayicon->setPixmap(langs[cur_lang]->pic);
}


bool KXNeurApp::x11EventFilter(XEvent *e)
{
    // let m_xkb process the event and emit signals if necessary
    xnkb->processEvent(e);
    return KApplication::x11EventFilter(e);
}

void KXNeurApp::groupChange(int groupno)
{
    if ( cur_lang != groupno ) {
	prev_lang = cur_lang;
	cur_lang = groupno;
	trayicon->setPixmap(langs[cur_lang]->pic);
    }
}


void KXNeurApp::refreshLang()
{
    QStringList lngs;
    xnkb->getGroupNames(lngs);
    QString lng;
    int i = 0;


    for ( ; i < MAX_LANGUAGES ; i++ ) {
	if ( langs[i] != NULL ) {
	    trayicon->menu->removeItem(langs[i]->menuid);
	    delete langs[i];
	}
	langs[i] = NULL;
    }

    i = 0;
    for ( QStringList::Iterator iter = lngs.begin(); iter != lngs.end() ; iter++, i++ ) {
	if ( i >= MAX_LANGUAGES ) {
	    qDebug("Error: found more then 4 langs!!!");
	    xneur_stop();
	    exit(1);
	}
	lng = *iter;
	QString code = all_langs->readEntry(lng);
	if ( code == QString::null ) {
	    langs[i] = new XNLang(i18n("* Unknown lang *"));
	    langs[i]->lg = QString::number(i);
	    langs[i]->supp_lg = -1;
	}
	else {
	    langs[i] = new XNLang(lng);
	    langs[i]->lg = code;
	    langs[i]->supp_lg = 0;
	}

	QPixmap pix(19, 16);
	QPainter painter(&pix);
	QFont font("helvetica", 9, QFont::Bold);
	font.setPixelSize(10);
	painter.setFont(font);
	if ( langs[i]->supp_lg >= 0 ) {
            painter.setPen(KGlobalSettings::highlightedTextColor());
            pix.fill(KGlobalSettings::highlightColor());
	    painter.drawText(1, 0, pix.width(), pix.height(), Qt::AlignHCenter | Qt::AlignVCenter, langs[i]->lg);
	}
	else {
	    pix.fill();
	    painter.drawText(0, 0, pix.width(), pix.height(), Qt::AlignHCenter | Qt::AlignVCenter, langs[i]->lg);
	}
	langs[i]->pic = pix;
    };

    for ( int j = 0 ; j < xnconf->total_languages ; j++ ) {
	QString code = all_langs->readEntry(xnconf->get_lang_name(xnconf, j));
	int l = -1;
	if ( code != QString::null ) {
	    for ( i = 0 ; i < MAX_LANGUAGES && langs[i] != NULL ; i++ ) {
		if ( code == langs[i]->lg ) {
		    langs[i]->name = xnconf->get_lang_name(xnconf, j);
		    langs[i]->supp_lg = 1;
		    l = i;
		    break;
		}
	    }
	    if ( l >= 0 ) {
		QString path = locate("appdata", langs[l]->name+".png");
		QPixmap pic;
		if ( KXNeurSettings::ShowInTray() == SHOW_ICON && !path.isEmpty() ) {
			// printf("show flag\n");
		    pic.load(path);
		    langs[l]->pic = pic;
		}
	    }
	}
    }

    for ( i = 0 ; i < MAX_LANGUAGES && langs[i] != NULL ; i++ )
	langs[i]->menuid = trayicon->menu->insertItem(langs[i]->pic, langs[i]->name, i, i+1);
}


void KXNeurApp::setNextLang()
{
    int groupno;
    if ( KXNeurSettings::SwitcherMode() && prev_lang >= 0 ) {
	groupno = prev_lang;
	// qDebug("switch mode");
    }
    else {
        groupno = cur_lang + 1;
	if ( groupno >= MAX_LANGUAGES || langs[groupno] == NULL )
	    groupno = 0;
	// qDebug("ring mode");
    }
    xnkb->setGroupNo(groupno);
}


void KXNeurApp::setMenuLang(int nn)
{
    // qDebug("click lang %d\n", nn);
    for ( uint i = 0 ; i < MAX_LANGUAGES && langs[i] != NULL ; i++ )
	if ( langs[i]->menuid == nn ) {
	    xnkb->setGroupNo(i);
	    break;
	}
}


XNLang::XNLang(const QString& n)
{
    name = n;
}

XNLang::~XNLang()
{
}
